### PR TITLE
# 1 Initial implementation

### LINKED ISSUE #1: Initial implementation

### PR DESCRIPTION
**Pull request by jakobbraun**
Closes #1

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Virtual Schema for Document Files in BucketFS
```
Comment (morazow) [inline README.md:1]:
```suggestion
# Virtual Schema for Document Files in S3
```

File: doc/user_guide/user_guide.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+Now you need to define a connection that includes the location of stored files:
+
+ ```
+CREATE CONNECTION S3_CONNECTION
+    TO 'AWS URI'
+    USER 'AWS ACCESS KEY'
+    IDENTIFIED BY 'AWS SECRET KEY';
+``` 
+
+The address (`TO`) must have one of theses formats:
```
Comment (morazow) [inline doc/user_guide/user_guide.md:47]:
```suggestion
The address (`TO`) must have one of these formats:
```

File: doc/user_guide/user_guide.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+## Defining the Schema Mapping
+
+Before creating a Virtual Schema you need to create a mapping definition that defines how the document data is mapped to Exasol tables.
+
+For that we use the Exasol Document Mapping Language (EDML). It is universal over all document Virtual Schemas. 
+To learn how to  define such EDML definitions check the [user guide in the common repository for all document Virtual Schemas](https://github.com/exasol/virtual-schema-common-document/doc/user_guide/edml_user_guide.md).
+
+In the definitions you have to define the `source` property. 
+For S3, you there define the key of the files to load.
```
Comment (morazow) [inline doc/user_guide/user_guide.md:67]:
```suggestion
For S3, you define the key of the files to load.
```

Comment (morazow) [reply doc/user_guide/user_guide.md:67]: What is meant by `key of the files`? Is it the names of the files?
Also here short example from the EDML mapping header (including source property) would help.

File: src/main/java/com/exasol/adapter/document/files/S3Uri.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+     * </p>
+     * 
+     * @param url URI to parse.
+     * @return built {@link S3Uri}
+     */
+    public static S3Uri fromString(final String url) {
+        final Matcher matcher = URL_PATTERN.matcher(url);
+        if (!matcher.matches()) {
+            throw new IllegalArgumentException(
+                    "The given S3 Bucket string has an invalid format. Expected Format: http(s)://BUCKET.s3.REGION.amazonaws.com/KEY or http(s)://BUCKET.s3.REGION.CUSTOM_ENDPOINT/KEY.");
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/files/S3Uri.java:41]:
```suggestion
                    "The given S3 Bucket string has an invalid format. Expected format: http(s)://BUCKET.s3.REGION.amazonaws.com/KEY or http(s)://BUCKET.s3.REGION.CUSTOM_ENDPOINT/KEY.");
```

File: src/test/java/com/exasol/adapter/document/files/S3DocumentFilesAdapterIT.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                        .create(LOCAL_STACK_CONTAINER.getAccessKey(), LOCAL_STACK_CONTAINER.getSecretKey())))
+                .region(Region.of(LOCAL_STACK_CONTAINER.getRegion())).build();
+
+        s3.createBucket(builder -> builder.bucket(TEST_BUCKET));
+        s3.putObject(builder -> builder.bucket(TEST_BUCKET).key("testData-1.json"),
+                RequestBody.fromBytes("{\"id\": \"book-1\"}".getBytes()));
+        s3.putObject(builder -> builder.bucket(TEST_BUCKET).key("testData-2.json"),
+                RequestBody.fromBytes("{\"id\": \"book-2\"}".getBytes()));
+        s3.putObject(builder -> builder.bucket(TEST_BUCKET).key("otherData.json"),
+                RequestBody.fromBytes("{\"id\": \"other\"}".getBytes()));
```
Comment (morazow) [inline src/test/java/com/exasol/adapter/document/files/S3DocumentFilesAdapterIT.java:88]: Put them into a S3 preparation method.

**Review by morazow — CHANGES_REQUESTED** [review]

File: doc/user_guide/user_guide.md (hunk 3 above)
Comment (jakobbraun) [inline doc/user_guide/user_guide.md:67]:
```suggestion
For S3, you there define the key (S3 term for file identifier) of the files to load.
```

Comment (jakobbraun) [reply doc/user_guide/user_guide.md:67]: see https://github.com/exasol/s3-document-files-virtual-schema/pull/2#discussion_r494982707

Comment (morazow) [reply doc/user_guide/user_guide.md:67]: This is okay, please commit it, but do not forget to remove the `there`.

File: src/test/java/com/exasol/adapter/document/files/S3DocumentFilesAdapterIT.java (hunk 5 above)
Comment (morazow) [reply src/test/java/com/exasol/adapter/document/files/S3DocumentFilesAdapterIT.java:97]: :+1:

**Review by morazow — APPROVED** [review]

**Pull request merged**